networkx/algorithms/polynomials/tutte.py

```diff
@@ -0,0 +1,180 @@
+    from `G` by contracting edge `e`, `k(G)` the number of cut-edges of `G`,
+    and `l(G)` the number of loops of `G`:
+
+    .. math::
+        T_G(x, y) = \begin{cases}
+    	   x^{k(G)} y^{l(G)}, & \text{if all edges are cut-edges or loops} \\
+           T_{G-e}(x, y) + T_{G/e}(x, y), & \text{otherwise, for $e$ not a cut-edge or loop}
+        \end{cases}
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    simplify : bool
+        determines if the returned expression will be in simplified form
+
+    Returns
+    -------
+    instance of :class:`sympy.core.add.Add`
+        A Sympy expression representing the Tutte polynomial for `G`.
+
+    Examples
+    --------
+    >>> C = nx.cycle_graph(5)
+    >>> nx.tutte_polynomial(C)
+    x**4 + x**3 + x**2 + x + y
+
+    >>> D = nx.diamond_graph()
+    >>> nx.tutte_polynomial(D)
+    x**3 + 2*x**2 + 2*x*y + x + y**2 + y
+
+    Notes
+    -----
+    This function implements the deletion-contraction recurrence (Definition 3,
+    described above). Edge contraction is defined and deletion-contraction is
+    introduced in [3]_. Combinatorial meaning of the coefficients is introduced
+    in [4]_. Universality, properties, and applications are discussed in [5]_.
+
+    See Also
+    --------
+    _get_cut_edges
+
+    References
+    ----------
+    .. [1] Y. Shi, M. Dehmer, X. Li, I. Gutman,
+       "Graph Polynomials," p. 14
+    .. [2] Y. Shi, M. Dehmer, X. Li, I. Gutman,
+       "Graph Polynomials," p. 46
+    .. [3] D. B. West,
+       "Introduction to Graph Theory," p. 84
+    .. [4] G. Coutinho,
+       "A brief introduction to the Tutte polynomial"
+       Structural Analysis of Complex Networks, 2011
+       https://homepages.dcc.ufmg.br/~gabriel/seminars/coutinho_tuttepolynomial_seminar.pdf
+    .. [5] J. A. Ellis-Monaghan, C. Merino,
+       "Graph polynomials and their applications I: The Tutte polynomial"
+       Structural Analysis of Complex Networks, 2011
+       https://arxiv.org/pdf/0803.3079.pdf
+    """
+    import sympy
+
+    G = nx.MultiGraph(G)
+
+    components = list(nx.connected_components(G))
+    if len(components) > 1:
+        t = 1
+        for component in components:
+            t *= tutte_polynomial(G.subgraph(component), False)
+        return sympy.simplify(t)
+
+    cut_edges = _get_cut_edges(G)
+    loops = [e for e in G.edges if e[0] == e[1]]
+    edges_not_cuts_loops = [i for i in G.edges if i not in cut_edges and i not in loops]
+
+    if not edges_not_cuts_loops:
+        return sympy.Symbol("x") ** len(cut_edges) * sympy.Symbol("y") ** len(loops)
+
+    e = edges_not_cuts_loops[0]
+    # deletion-contraction
+    C = nx.contracted_edge(G, e, self_loops=True)
+    contraction_loops = [i for i in C.edges if i[0] == i[1] and i not in G.edges]
+    if contraction_loops:
+        C.remove_edge(*contraction_loops[0])
+    G.remove_edge(*e)
+
+    t = tutte_polynomial(G, False) + tutte_polynomial(C, False)
```

Comment:
How big of a graph are you expecting this to work with? 
This recursion is likely to cause memory problems and run into recursion limits for graphs with many hundreds of nodes.  Can you do it without recursion?

Reply:
@dschult  @rossbar This is a good point and a known challenge. The recursive deletion-contraction method is the standard (and, until recently, fastest) general approach. Retrieving the coefficients is [#P-complete](https://dl.acm.org/doi/abs/10.1016/0166-218X(94)00097-W).

The fastest result I am aware of computes the coefficients in [vertex-exponential time](https://arxiv.org/abs/0711.2585). My intention was, if this feature were accepted as useful or interesting, to try implementing the vertex-exponential method as a companion function (similar to `gnp_random_graph` vs. `fast_gnp_random_graph`).

Reply:
Hmmm   I don't think recursive should impact the proof of speed of the algorithm. Any recursive algorithm can be rewritten as an iterative algorithm with a stack -- indeed that is what the programming language is doing in the background when you use recursion.  There is a stack of function over-heads that gets added to when another call is made and deleted from when each function call returns. 

The idea behind writing at as iterative (and one reason Python puts an adjustable recursion limit on it's kernels) is because the function calls are not a very good way to make and use a stack. Recursion is good for proving theorems, but it often is slow and a memory hog. 

I'm fine with including a recursive version if it is too hard to make one that is iterative. But it will limit the size of graphs, so I wondered what size you expect people to use with this algorithm?

Reply:
And "yes", it would be cool to write the other algorithm -- though if it is both fast and memory efficient, maybe it's best to make it the main function with the recursive version available for pedagogical value.

Reply:
@dschult Thank you for your careful review and explanation. I see now what you mean about writing at as iterative - I had misunderstood.

I have replaced the algorithm with a more practical iterative approach and relegated the recursive implementation to `tutte_polynomial_recursive` for pedagogical purposes. Is this preferable?

Thanks again!

Reply:
@dschult  Thank you for your careful review and explanation. I see now what you mean about rewriting as iterative - I had misunderstood.

> The idea behind writing at as iterative (and one reason Python puts an adjustable recursion limit on it's kernels) is because the function calls are not a very good way to make and use a stack. Recursion is good for proving theorems, but it often is slow and a memory hog.

This makes sense, thanks. In my most recent commit, `tutte_polynomial` is rewritten as an iterative algorithm with a stack, as you describe. The recursive implementation now appears in `tutte_polynomial_recursive` for pedagogical purposes.

> I'm fine with including a recursive version if it is too hard to make one that is iterative. But it will limit the size of graphs, so I wondered what size you expect people to use with this algorithm?

I don't expect this to be used for graphs with more than, say, a dozen vertices. More than that would take several hours, and even the most performant general implementation I'm aware of indicates a limit of 25 vertices ([Section D.2](https://arxiv.org/pdf/0711.2585.pdf)).

Luckily, as far as I'm aware, the graphs of interest for practical computations are small, e.g. for [biomolecular computing](https://link.springer.com/chapter/10.1007/978-3-319-08019-2_7).